### ISSUE TITLE
The Baseline Roadmap

### ISSUE THREAD

**Issue by Kulbear** [Kulbear/QuickDraw#1]
- [x] Play with the starter kit: [LB 0.892](https://www.kaggle.com/gaborfodor/greyscale-mobilenet-lb-0-892/notebook)
- [x] Dataset and DataLoader implementation
    - [x] Augmentation
    - [x] Generator implementation of DataLoader
- [x] Adapt SE-ResNet and SE-InceptionV4
- [x] Implement flipping TTA

**Comment by Kulbear** [comment]
## Doodle Classification Note

Training config:
- ~ 75% of 30k / class ( like ~23k / class)
- size 128 x 128
- batch size 128 (can be a bit more larger)
- 500, steps every epoch, so every epoch train with 128 * 500 images
- Adam, lr start at 2e-4, use cyclic LR with cos annealing 
```python
def cos_annealing_lr(initial_lr, cur_epoch, epoch_per_cycle):
    return initial_lr * (np.cos(np.pi * cur_epoch / epoch_per_cycle) + 1) / 2
```
- 6 cycles, each cycle 25 epochs. By checking the training log I think this schedule should be improved further)

Model used (with best score):
- MobileNet 0.913 (by hflip TTA)
- ResNet18 0.919
- Xception 0.925 (not with the current 3 channel information, I think the current one can give a better result)
- SE-ResNeXt50 (32 x 4d) 0.930

Ideas used: 
- imagenet-pretrained weights
- larger images
    I tried Xception with 224 x 224 image, by using about 20k/class, and just ReduceOnPlaque can get 0.925, I think larger --> better
- grayscale strokes to ordered strokes
- country code information encoding 
```python
c_value = country_mapping.get(df.iloc[i]['countrycode'], 40)
x[i, :, :, 0] = img + mask * c_value
```
- highlight first and last stokes (may drop the last and retry)
```python
if i == 0 or i == len(stroke[0]) - 2:
    color = 255
    _ = cv2.line(img, (stroke[0][i], stroke[1][i]),
                 (stroke[0][i + 1], stroke[1][i + 1]), color, lw)
```
- hflip TTA (good when LB < 0.920, bad when LB >= 0.920)
- last few layers in the model, with dropout rate 0.075 (prob not good)
```python
    def forward(self, x):
        x = self.features(x)
        x = self.relu(x)
        x = self.avg_pool(x)
        x = x.view(x.size(0), -1)
        x = self.dropout(x)
        x = self.last_linear(x)
```

Possible Models:

Size 1:

- Xception
- ResNet50
- SE ResNeXt50

Size 2:
- SE ResNeXt101
- DPN?
- InceptionResV2


TODO:
- better LR scheduler (this seems to be the most important one)
- update network after several batches to increase the valid batch size
- finetune with different optimizers?
- larger batches
- lstm
    - bi-lstm
    - gru?
    - cnn as feature extractor and rnn accept result from global pooling?
- label denoising

**Comment by Kulbear** [comment]
![11_203_h3_1_blue](https://user-images.githubusercontent.com/14886380/49769530-c762b500-fc9d-11e8-8580-e881e1605458.png)
![11_203_h3_1_green](https://user-images.githubusercontent.com/14886380/49769531-c762b500-fc9d-11e8-9a68-55739922dacc.png)
![11_203_h3_1_red](https://user-images.githubusercontent.com/14886380/49769532-c762b500-fc9d-11e8-9598-4fdd9d32a4ad.png)
![11_203_h3_1_yellow](https://user-images.githubusercontent.com/14886380/49769533-c762b500-fc9d-11e8-9570-fbcfaf4ca370.png)

**Comment by Kulbear** [comment]
https://www.kaggle.com/c/human-protein-atlas-image-classification/discussion/74374#437548
```
train_df_orig=train_df.copy()    
lows = [15,15,15,8,9,10,8,9,10,8,9,10,17,20,24,26,15,27,15,20,24,17,8,15,27,27,27]
for i in lows:
    target = str(i)
    indicies = train_df_orig.loc[train_df_orig['Target'] == target].index
    train_df = pd.concat([train_df,train_df_orig.loc[indicies]], ignore_index=True)
    indicies = train_df_orig.loc[train_df_orig['Target'].str.startswith(target+" ")].index
    train_df = pd.concat([train_df,train_df_orig.loc[indicies]], ignore_index=True)
    indicies = train_df_orig.loc[train_df_orig['Target'].str.endswith(" "+target)].index
    train_df = pd.concat([train_df,train_df_orig.loc[indicies]], ignore_index=True)
    indicies = train_df_orig.loc[train_df_orig['Target'].str.contains(" "+target+" ")].index
    train_df = pd.concat([train_df,train_df_orig.loc[indicies]], ignore_index=True)
```
